### PR TITLE
feat:train simple-ddpm after training auto-encoder

### PR DESCRIPTION
**Pull request by jacky-chen-dev**
# Draft Changelog
Last update: 10 Sep 2024 6:22 pm

## Added
- adapted a simple DDPM from https://github.com/cloneofsimo/minDiffusion
  - modify its input shape to [batch_size, input_dim], and the corresponding noise/beta shape
  - replace conv layers with linear layers, so to lower complexity
- add `insert_ddpms` function in ContactGen model
- add the `train_ddpm` function in `trainer.py`
  -  after training the auto-encoder, record the fixed latent space
  - use samples from the latent space to train 3 DDPM, the samples are detached so as to prevent the backpropgation to the mean and std of normal distribution
  - the training loop of `train_ddpm` function are adapted from https://github.com/cloneofsimo/minDiffusion/blob/master/superminddpm.py

## Result
Trained auto-encoder for 10 epochs, then insert and train the ddpm.

the loss increased rapidly after inserting DDPM (the 10th epoch),  then drop to a minimum, finally the loss increases again

I will try to tune the model to decrease the loss in future pr.
螢幕截圖 2024-09-10 上午12 28 39

### CODE REVIEW

**Comment by Yiyao-Ma** [comment]
For this bug, adding `retain_graph=True` in `loss.backward()` can solve it.
Please refer this [link](https://stackoverflow.com/questions/48274929/pytorch-runtimeerror-trying-to-backward-through-the-graph-a-second-time-but).

**Comment by jacky-chen-dev** [comment]
The changelog is updated. This will be the last pr in this repo. I will create a private repository this week and migrate to there.
